include/nonunity/interfaces/NEML2ModelInterface.h

```diff
@@ -0,0 +1,124 @@
+/****************************************************************/
+
+#pragma once
+
+#include "neml2/models/Model.h"
+#include "neml2/misc/parser_utils.h"
+#include "Material.h"
+#include "UserObject.h"
+
+/**
+ * Interface class to provide common input parameters, members, and methods for MOOSEObjects that
+ * use NEML2 models.
+ */
+template <class T>
+class NEML2ModelInterface : public T
+{
+public:
+  static InputParameters validParams();
+
+  template <typename... P>
+  NEML2ModelInterface(const InputParameters & params, P &&... args);
+
+protected:
+  /**
+   * Validate the NEML2 material model. This method should throw a moose error with the first
+   * encountered problem. Note that the developer is responsible for calling this method at the
+   * appropriate times, for example, at initialSetup().
+   */
+  virtual void validateModel() const;
+
+  /// Get the NEML2 model
+  neml2::Model & model() const { return _model; }
+
+  /// Get the target compute device
+  const torch::Device & device() const { return _device; }
+
+  /**
+   * @brief Convert a raw string to a LabeledAxisAccessor
+   *
+   * @param raw_str
+   * @return neml2::LabeledAxisAccessor
+   */
+  neml2::LabeledAxisAccessor getLabeledAxisAccessor(const std::string & raw_str) const;
+
+private:
+  /// The NEML2 material model
+  neml2::Model & _model;
+
+  /// The device on which to evaluate the NEML2 model
+  const torch::Device _device;
+};
+
+template <class T>
+InputParameters
+NEML2ModelInterface<T>::validParams()
+{
+  InputParameters params = T::validParams();
+  params.addRequiredParam<std::string>(
+      "model",
+      "Name of the NEML2 model, i.e., the string inside the brackets [] in the NEML2 input file "
+      "that corresponds to the model you want to use.");
+  params.addParam<std::string>(
+      "device",
+      "cpu",
+      "Device on which to evaluate the NEML2 model. The string supplied must follow the following "
+      "schema: (cpu|cuda)[:<device-index>] where cpu or cuda specifies the device type, and "
+      ":<device-index> optionally specifies a device index. For example, device='cpu' sets the "
+      "target compute device to be CPU, and device='cuda:1' sets the target compute device to be "
+      "CUDA with device ID 1.");
+  return params;
+}
+
+template <class T>
+template <typename... P>
+NEML2ModelInterface<T>::NEML2ModelInterface(const InputParameters & params, P &&... args)
+  : T(params, args...),
+    _model(neml2::Factory::get_object<neml2::Model>("Models", params.get<std::string>("model"))),
+    _device(params.get<std::string>("device"))
+{
+  // Send the model to the compute device
+  _model.to(_device);
+}
+
+template <class T>
+void
+NEML2ModelInterface<T>::validateModel() const
+{
+  // Forces and old forces on the input axis must match, i.e. all the variables on the old_forces
+  // subaxis must also exist on the forces subaxis:
+  if (_model.input().has_subaxis("old_forces"))
```

Comment:
If I understand this right, this class is a generic interface. Isn't 'force' a mechanics-specific variable? If that's the case, would this check belong in `NEML2SolidMechanicsInterface`?

Reply:
Here "force" is more of a thermodynamic force rather than a mechanical force. See the naming conventions section here for more details: https://reverendbedford.github.io/neml2/devel.html

But I surely agree with you that this is confusing. Mark and I had an internal debate at the very beginning of the NEML2 development, and unfortunately I think now is a little too late to change that naming convention.